### ISSUE TITLE
Normalmap inverts at certain angles on mesh rotated by bone.

### ISSUE THREAD

**Issue by jitspoe** [godotengine/godot#81446]
### Godot version

4.1.1.stable.official

### System information

Windows 10, Vulkan forward +

### Issue description

Under certain conditions, it's possible for the normal map to effectively invert the lighting, as seen here:
![revolver_normal](https://github.com/godotengine/godot/assets/7751182/c0ab812a-c6fd-4d89-bbfd-8dc409f176fb)

I tried to make some simpler repro cases in the background, but, for some reason, the issue only happens with my revolver.  This was originally happening with a procedurally rotated bone using skeleton.set_bone_pose_rotation, but I tested with an animation driving the bone and the issue still happened.

My hypothesis is that the tangent and/or binormal are getting miscalculated because a cross product is attempted with 2 vectors that are the same somehow, but I could be way off.

### Steps to reproduce

I attempted to recreate this using a simple skinned 6-sided cylinder mesh with a rotation animation by itself, but, alas, the issue only seems to repro with my KOOK game asset, so I've included that in the repro project:

### Minimal reproduction project

[test_animation_normalmap.zip](https://github.com/godotengine/godot/files/12558137/test_animation_normalmap.zip)

**Comment by bitsawer** [comment]
I can confirm this on current master too (0bca4242392992840b0b891acde6422c49712e3e). Pretty interesting, I would also guess that something goes wrong with the skinning tangent space or related calculation when a bone is rotated/scaled in a certain way. Manually rotating `rr_cyl1` bone around its cylider axis in pose mode also seems to cause this bug, but if you also rotate bone's another axis a bit it seems to "fix" the issue, so there could indeed be some kind of axis issue.

**Comment by jitspoe** [comment]
Yeah, certainly bizarre that I couldn't seem to reproduce it with basically the same kind of mesh as well, just isolated to a 6-sided cylinder.  I added like 0.01 rotation on another axis to work around it for now.  Thanks for the temp fix idea.

I thought maybe it was something to do with the GLTF tangents, as I had another asset that had inverted normals and I had to shut off tangent exports and let Godot handle the tangents, but I also tried that on this asset and it didn't fix it.

**Comment by jitspoe** [comment]
So this just happened on another model with a similar setup.  Once again, adding a slight rotation on another axis fixed it.  Seems the issue still occurs in 4.3, in case anybody was wondering.

**Comment by jitspoe** [comment]
It seems this issue goes beyond just normal maps.  The vertex normals themselves seem to be broken even with no texture.  This is in the import preview window:

<img width="636" height="613" alt="Image" src="https://github.com/user-attachments/assets/9822c1a8-814b-4a5e-98f8-c1b833f9241a" />

Notice how one cylinder has the normals on top like blended with the normals in front.  I thought it might be a compression issue, but I checked "Force disable compression" and it still occurs.

**Comment by Calinou** [comment]
Do you have Generate LODs enabled on the scene? Try disabling it and see if the issue persists.

Also, did you triangulate your mesh on export in Blender (either by checking the appropriate option in the glTF export dialog, or adding a Triangulate modifier and ensuring modifiers are applied on export)? This is recommended, so that Blender and Godot use the same triangulation pattern.

**Comment by jitspoe** [comment]
Tried disabling LOD, but that made no difference.  I'm not sure it would make a difference for the preview window anyway, because doesn't that show the full LOD version?  Also, the normals break and un-break based on the bone rotations, so I don't think it's specifically an issue with the import so much as something with how the shader handles normals or the way the skeleton transforms them, maybe?  I couldn't find an option to triangulate with .blend/.gltf.  It shouldn't matter in this case, because the cylinders are flat shaded and flat quads, so regardless of the triangulation, the quads should light the same.  This is beyond a simple case like that, though.  The normals are completely broken in some cases.

I did, however try .fbx, and that seemed to not exhibit the issue, though I haven't tested it animating (sometimes it comes and goes depending on the rotation of the bones).

Edit: If somebody wants to do a diff on the meshes imported with FBX vs GLTF, I'd be really curious why one works properly and the other doesn't.

[revolutionary_revolver.zip](https://github.com/user-attachments/files/22671725/revolutionary_revolver.zip)

**Comment by jitspoe** [comment]
Has this been tested with my revolver?  I still have the issue in 4.7, and I can't reopen this.

**Comment by Gaktan** [comment]
Indeed, my PR only fixed a part of the issue, but there was still some edge cases. I found a better fix

**Comment by jitspoe** [comment]
Commenting here as well: Any idea why this issue is only occurring for me with .blend/gltf import, and not with FBX?  Also, if I import with force disable compression on the .blend file, I still get the problem.  The attempted fixes seem to be related to compression in the shader.  I'm wondering if this is a different issue.

**Comment by paddy-exe** [comment]
Reopened due to incompleteness (see https://github.com/godotengine/godot/issues/81446#issuecomment-4878103661)

**Comment by Gaktan** [comment]
> Commenting here as well: Any idea why this issue is only occurring for me with .blend/gltf import, and not with FBX? Also, if I import with force disable compression on the .blend file, I still get the problem. The attempted fixes seem to be related to compression in the shader. I'm wondering if this is a different issue.

Yes, there were two separate bugs, but closely related to each other. The first issue that I fixed was and edge case where a specific tangent direction would cause the tangent vector to be interpreted as "compressed" rather than uncompressed. Basically the tangent would be decoded as random garbage. I think the issue is shown here: https://github.com/godotengine/godot/issues/81446#issuecomment-3287599463.

The second issue that is visible in the first comment is caused by floating point precision issues that seem to "randomly" flip the sign of the bitangent, causing normal maps to flip direction.

I don't know why this doesn't happen with FBX though. Maybe tangents are recomputed in one format, but not the other. Maybe it's a small floating point precision mismatch between the two. Basically your issue occurs when a skinned mesh has tangent vectors that are pointing at a very specific axis-aligned direction. Any nudge in any direction and the issue would not happen essentially.
